src/protocols/secure_channel/SessionResumptionStorage.cpp

```diff
@@ -0,0 +1,189 @@
+/*
+ *    Copyright (c) 2022 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines the CHIP CASE Session object that provides
+ *      APIs for constructing a secure session using a certificate from the device's
+ *      operational credentials.
+ */
+
+#include <protocols/secure_channel/SessionResumptionStorage.h>
+
+#include <lib/support/Base64.h>
+#include <lib/support/SafeInt.h>
+
+namespace chip {
+
+constexpr TLV::Tag SessionResumptionStorage::kFabricIndexTag;
+constexpr TLV::Tag SessionResumptionStorage::kPeerNodeIdTag;
+constexpr TLV::Tag SessionResumptionStorage::kResumptionIdTag;
+constexpr TLV::Tag SessionResumptionStorage::kSharedSecretTag;
+constexpr TLV::Tag SessionResumptionStorage::kCATTag;
+
+const char * SessionResumptionStorage::StorageKey(DefaultStorageKeyAllocator & keyAlloc, ScopedNodeId node)
+{
+    return keyAlloc.FabricSession(node.GetFabricIndex(), node.GetNodeId());
+}
+
+const char * SessionResumptionStorage::StorageKey(DefaultStorageKeyAllocator & keyAlloc, ConstResumptionIdView resumptionId)
+{
+    char resumptionIdBase64[(resumptionId.size() + 2) / 3 * 4 + 1];
+    Base64Encode(resumptionId.data(), resumptionId.size(), resumptionIdBase64);
+    return keyAlloc.SessionResumption(resumptionIdBase64);
+}
+
+CHIP_ERROR SessionResumptionStorage::Save(ScopedNodeId node, ConstResumptionIdView resumptionId,
+                                          const Crypto::P256ECDHDerivedSecret & sharedSecret, const CATValues & peerCATs)
+{
+    DefaultStorageKeyAllocator keyAlloc;
+    {
+        // Save session state into key: /f/<fabricIndex>/s/<nodeId>
```

Comment:
It does not appear that storage for this feature is bounded.  I would think that to use this in production, we would need some type of mechanism (fifo?) to delete old entries.

It might also be nice to have some possibility to prune entries on fabric deletion, although I assume we would fall back to normal CASE if there was an attempt to use resumption data from a deleted fabric.

Reply:
I am pointing out that the potential number of peers a node may encounter over its lifetime is unbounded and, so with the current kvstore key pattern, `/f/<fabricIndex>/s/<nodeId>`, our storage requirement is up to 255 * 0xFFFF_FFEF_FFFF_FFFE entries for this feature.

So a node using this feature may slowly fill up storage over its lifetime with these entries, and there's nothing to stop that from growing over time.

Plausible options:

* Specify that this feature requires injection of a special storage delegate that can be bounded in size and can drop entries when it is too full; implement such a special storage delegate for an example app.
* Change the kvstore key pattern to a hash of fabric-index+nodeId with a mapping to a small, fixed number of entries corresponding to a compile-time-defined number of storage entries the platform can allocate to this feature.

And maybe there are other solutions too.  But we can't just fill up storage without any limit.  This will result in some nodes slowly degrading over time until they are unable to function.

Reply:
+1 we have to have storage bounds, and very small ones for a start